### PR TITLE
feat: implement creating a conversation

### PR DESCRIPTION
**Pull request by JChris246**
### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/BoopChat/boop/pulls) the bugtracker for similar pull requests
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes

---

### What is the purpose of your *pull request*?
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Description of your *pull request* and other information

This PR adds functionality to the Conversation component to be able to tell the server to create a conversation with the given participant list.

## Screenshots (if appropriate):
DeepinScreenshot_20210920022420

### CODE REVIEW

File: client/src/components/messenger/Conversations.js [hunk 1]
```diff
@@ -1,53 +1,112 @@
...
+    useEffect(() => {
+        if (!init) { // if this is the first time rendering, get user contacts from server
+            (async () => setContacts(await ContactsController.getContacts( token )))();
+            setInit(true);
+        }
+    }, [init]);
+
+    return (
+        <Dialog onClose={handleClose} aria-labelledby="simple-dialog-title" open={open}>
+            <DialogTitle id="simple-dialog-title">Create a conversation</DialogTitle>
```
Comment (krazeelazy) [inline client/src/components/messenger/Conversations.js:31]: Shouldn't the user be allowed to put in a title as well?

File: client/src/components/messenger/Conversations.js [hunk 2]
```diff
@@ -1,53 +1,112 @@
...
             <div id="chats">
                 {conversations.map((chat, i) =>
                     <ConversationItem
-                        name={chat.name}
-                        img={chat.img}
+                        name={chat.title}
+                        img={chat.imgUrl}
                         lastMsg={chat.lastMsg}
-                        lastDate={chat.lastDate}
+                        lastDate={ConversationsController.evaluateDate(chat.lastDate)}
```
Comment (krazeelazy) [inline client/src/components/messenger/Conversations.js:101]: Where does lastDate come from?

Comment (JChris246) [reply client/src/components/messenger/Conversations.js:101]: I believe this is the creation date of the last message in that conversation

Comment (krazeelazy) [reply client/src/components/messenger/Conversations.js:101]: Yh I get that but where is it assigned? Same with lastMsg

File: client/src/components/messenger/Conversations.js [hunk 3]
```diff
@@ -1,53 +1,112 @@
...
+    const [conversations, setConversations] = useState([]);
+    const [init, setInit] = useState(false);
+
+    // Get the token from the users global state.
+    const token = useSelector((state) => state.user.token);
+
+    useEffect(() => {
+        // if this is the first time rendering, get user conversations from server
+        if (!init) {
+            (async () => setConversations(await ConversationsController.getConversations(token)))();
```
Comment (davidfowl) [inline client/src/components/messenger/Conversations.js:63]: I know why you're doing this immediately executing async lambda, but I don't love it 😄

File: client/src/components/messenger/Chat.js [hunk 4]
```diff
@@ -1,54 +1,47 @@
-import React from "react";
+import {useState, useEffect, React} from "react";
```
Comment (davidfowl) [inline client/src/components/messenger/Chat.js:1]: nit: spacing

File: client/src/components/messenger/Contacts.js [hunk 5]
```diff
@@ -60,34 +40,23 @@ const Contacts = () => {
...
-
-                const result = await res.json();
-                if (res.status !== 201)
-                    alert(result.message); // contact was not added ... display why
-                else // add new contact to the back of the list
+                let result = ContactsController.addContact(token, email);
+                if (result.success) // add new contact to the back of the list
                     setContacts(contacts.length > 0 ? [...contacts, result.contact]: [result.contact]);
+                else // display error message
+                    alert(result.msg);
```
Comment (davidfowl) [inline client/src/components/messenger/Contacts.js:59]: Why alert?

File: server/controllers/Conversations.js [hunk 6]
```diff
@@ -209,6 +216,9 @@ module.exports.leaveConversation = async (req, res) => {
     }
 
     if(userIsAdmin == true){
```
Comment (davidfowl) [inline server/controllers/Conversations.js:218]:
```suggestion
    if (userIsAdmin === true) {
```

File: client/src/components/messenger/controllers/Contacts.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        // get the list of contacts if successful
+        return data ? data.contactList : [];
+    },
+    evaluateStatus: lastActive => {
+        // for now status is either online or offline
+        // if user not active within the last 5mins - offline
+        let diff = Date.now() - (new Date(lastActive)).getTime();
+        return diff > (5 * 60 * 1000) ? "offline": "online";
+    },
+    addContact: async ( token, email ) => {
```
Comment (davidfowl) [inline client/src/components/messenger/controllers/Contacts.js:21]: nit: Format document?

```suggestion
    addContact: async (token, email) => {
```

File: client/src/components/messenger/Contacts.js [hunk 8]
```diff
@@ -29,84 +31,80 @@ const AddContactDialog = ({open, onClose}) => {
...
-                        contactEmail: email
-                    })
-                });
-
-                const result = await res.json();
-                if (res.status !== 201)
-                    alert(result.message); // contact was not added ... display why
-                else // add new contact to the back of the list
+        if (email) {
+            const runAsync =async () => {
```
Comment (davidfowl) [inline client/src/components/messenger/Contacts.js:64]:
```suggestion
            const runAsync = async () => {
```

**Review by Pherboid — APPROVED** [review]

**Pull request merged**